### ISSUE TITLE
Cluster side channel blocks when multithreading

### ISSUE THREAD

**Issue by tanjiancheng** [JohnSully/KeyDB#102]
I use wrk tools to test the keydb cluster

```
[root@centos7 bin]# ./wrk -c500 -d10 -t4 http://10.11.81.174:8100/test/redis_set --latency   
Running 10s test @ http://10.11.81.174:8100/test/redis_set
  4 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    87.56ms   13.98ms  97.44ms  100.00%
    Req/Sec     8.75      8.35    20.00     37.50%
  Latency Distribution
     50%   97.44ms
     75%   97.44ms
     90%   97.44ms
     99%   97.44ms
  14 requests in 10.03s, 2.98KB read
  Socket errors: connect 0, read 0, write 0, timeout 12
Requests/sec:      1.40
Transfer/sec:     304.32B
```
when i benchmark the keydb cluster,it becomes very slow and It took a while to get back to normal , when i change the config to use redis cluster,and the same code,redis cluster is ok for response

```
[root@centos7 bin]# ./wrk -c500 -d10 -t4 http://10.11.81.174:8100/test/redis_set --latency 
Running 10s test @ http://10.11.81.174:8100/test/redis_set
  4 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    17.32ms   21.67ms 229.27ms   88.15%
    Req/Sec    10.73k     5.42k   19.81k    54.50%
  Latency Distribution
     50%    8.82ms
     75%   14.96ms
     90%   44.61ms
     99%  108.87ms
  427245 requests in 10.02s, 88.82MB read
Requests/sec:  42655.16
Transfer/sec:      8.87MB
```

here is my keydb-cluster-node config
```
port 38001
cluster-enabled yes
cluster-config-file nodes-38001.conf
cluster-node-timeout 5000
daemonize yes
protected-mode no
dir /data/KeyDB-5.1.0/nodes
server-thread-affinity true
server-threads 4
```
without server-thread-affinity and server-threads the result is the same,so can explan why happen? that

my keydb version is KeyDB-5.1.0, system is CentOS Linux release 7.6.1810 (Core)

thanks!

**Comment by JohnSully** [comment]
are you running multiple cluster nodes on the same system?  When you use server-thread-affinity KeyDB will bind each thread to cores sequentially. If you launch multiple instances with affinity set then they will have conflict over the same cores and will run slowly. 

With Redis it is common to setup clusters on the same machine to workaround its single threaded behavior.  This is not recommended for KeyDB. Each cluster node should be on its own machine. A single KeyDB instance is able to get the maximum performance out of your hardware.

**Comment by tanjiancheng** [comment]
> are you running multiple cluster nodes on the same system? When you use server-thread-affinity KeyDB will bind each thread to cores sequentially. If you launch multiple instances with affinity set then they will have conflict over the same cores and will run slowly.
> 
> With Redis it is common to setup clusters on the same machine to workaround its single threaded behavior. This is not recommended for KeyDB. Each cluster node should be on its own machine. A single KeyDB instance is able to get the maximum performance out of your hardware.

thanks for reply, I build the keydb cluster on two machine which has 3 master and 3 slave, so you means if I want to build 3 master and 3 slave cluseter I need to have 6 machine to do this? or have any way to bind the specified core by myself to use less machine ?

**Comment by tanjiancheng** [comment]
Excuse me， today I use the new three machines(not other cluster) to build the keydb cluster(just three master nodes on each machine) and test agin,the result is same,may be not the core bind problem
```
[root@centos7 bin]# ./wrk -c500 -d10 -t4 http://10.11.81.174:8100/test/redis_set --latency 
Running 10s test @ http://10.11.81.174:8100/test/redis_set
  4 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    39.60ms   11.09ms  47.45ms  100.00%
    Req/Sec    28.02     18.13    70.00     70.77%
  Latency Distribution
     50%   47.45ms
     75%   47.45ms
     90%   47.45ms
     99%   47.45ms
  386 requests in 10.03s, 82.18KB read
  Socket errors: connect 0, read 0, write 0, timeout 384
Requests/sec:     38.49
Transfer/sec:      8.19KB
```
here is keydb-node config
```
port 38000
cluster-enabled yes
cluster-config-file nodes-38000.conf
cluster-node-timeout 5000
daemonize yes
protected-mode no
dir /data/KeyDB-5.1.1/nodes
``` 

here is cluster nodes config
```
d3a5f5fa2843aa4309a221385088d6697ce57d63 10.11.81.179:38000@48000 master - 0 1572833590918 3 connected 10923-16383
246837749e2390f7d120cdd4a5b8f1bc086cf812 10.11.86.135:38000@48000 master - 0 1572833590816 1 connected 0-5460
47d2103274cfd98f46938680a3b689e3f27b5968 10.11.86.134:38000@48000 myself,master - 0 1572833589000 2 connected 5461-10922
vars currentEpoch 3 lastVoteEpoch 0
```
### cli info
```
127.0.0.1:38000> info
# Server
redis_version:5.1.1
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:88e0cd3ee4131527
redis_mode:cluster
os:Linux 3.10.0-123.el7.x86_64 x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:6.3.1
process_id:2911
run_id:39a3380bf20c9454919ae2368f58dfe921d069aa
tcp_port:38000
uptime_in_seconds:1598
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:12553208
executable:/data/KeyDB-5.1.1/src/keydb-server
config_file:/data/KeyDB-5.1.1/nodes/node-38000.conf

# Clients
connected_clients:3
client_recent_max_input_buffer:2
client_recent_max_output_buffer:0
blocked_clients:1
current_client_thread:0
thread_0_clients:3

# Memory
used_memory:9852192
used_memory_human:9.40M
used_memory_rss:38322176
used_memory_rss_human:36.55M
used_memory_peak:18319392
used_memory_peak_human:17.47M
used_memory_peak_perc:53.78%
used_memory_overhead:9806538
used_memory_startup:9720552
used_memory_dataset:45654
used_memory_dataset_perc:34.68%
allocator_allocated:10323312
allocator_active:10924032
allocator_resident:42274816
total_system_memory:6090694656
total_system_memory_human:5.67G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.06
allocator_frag_bytes:600720
allocator_rss_ratio:3.87
allocator_rss_bytes:31350784
rss_overhead_ratio:0.91
rss_overhead_bytes:-3952640
mem_fragmentation_ratio:3.89
mem_fragmentation_bytes:28471288
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_clients_slaves:0
mem_clients_normal:83850
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:0
rdb_changes_since_last_save:886
rdb_bgsave_in_progress:0
rdb_last_save_time:1572832698
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0

# Stats
total_connections_received:3189
total_commands_processed:4923
instantaneous_ops_per_sec:6
total_net_input_bytes:179835
total_net_output_bytes:17734134
instantaneous_input_kbps:0.21
instantaneous_output_kbps:19.11
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
evicted_keys:0
keyspace_hits:0
keyspace_misses:0
pubsub_channels:1
pubsub_patterns:0
latest_fork_usec:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

# Replication
role:master
connected_slaves:0
master_replid:e845cbbe65d70eb580ff411c4835548d759462fc
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:6.561721
used_cpu_user:4.724276
used_cpu_sys_children:0.000000
used_cpu_user_children:0.000000
server_threads:1
long_lock_waits:0

# Cluster
cluster_enabled:1

# Keyspace
db0:keys=1,expires=0,avg_ttl=0
```

### cli cluster info
```
127.0.0.1:38000> CLUSTER INFO
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:3
cluster_size:3
cluster_current_epoch:3
cluster_my_epoch:2
cluster_stats_messages_ping_sent:853
cluster_stats_messages_pong_sent:847
cluster_stats_messages_meet_sent:2
cluster_stats_messages_sent:1702
cluster_stats_messages_ping_received:847
cluster_stats_messages_pong_received:855
cluster_stats_messages_received:1702
```

**Comment by JohnSully** [comment]
If you would like to use multiple KeyDB instances on the same machine, then you cannot use the core binding flag as it will not coordinate between them.  We may fix this in the future.  One of the main goals of KeyDB was to eliminate the need to run multiple servers on the same machine, but you can still do so if you want to.

Most KeyDB users are using Active Replication, it's possible there may be some issue specific to clustering, I will attempt to repro this slowdown.

**Comment by tanjiancheng** [comment]
> If you would like to use multiple KeyDB instances on the same machine, then you cannot use the core binding flag as it will not coordinate between them. We may fix this in the future. One of the main goals of KeyDB was to eliminate the need to run multiple servers on the same machine, but you can still do so if you want to.
> 
> Most KeyDB users are using Active Replication, it's possible there may be some issue specific to clustering, I will attempt to repro this slowdown.

Ok，Thank you very much for your answer，why I want to use the keydb cluster because I test keydb is sure much faster than redis, but if I want to use mutiple machine to store data, I need to change my client or my code to make different key  distributed on different keydb instance,that is what cluster do by itself.I'm really looking forward to using keydb clusters in production

**Comment by JohnSully** [comment]
I've narrows this down the cluster side channel.  With the current implementation is will transmit and receive while holding the lock.  Not too difficult to fix.

**Comment by tanjiancheng** [comment]
Excuse me，Are there any plans for a release that will fix this problem?so I can test again

**Comment by JohnSully** [comment]
Hi @tanjiancheng 

What I previously thought was the problem is actually not responsible for what your seeing.  The core issue is that at moderate load latency is about 0.5ms higher with multi-threading.  This is not true at high load where we do most of our benchmarks.

The potential throughput of a KeyDB cluster with multi-threading is extremely high, I wasn't able to saturate it with a 64-core server running memtier.  As a result these cluster benchmarks are essentially testing latency at low to moderate loads.

KeyDB should not have higher latency (even if it's only 0.5ms) in this scenario so I have a plan to fix that but it's quite complex.  Essentially I have to write a load balancer for clients.  However because the actual throughput is limited by the benchmark tool and not KeyDB the best this will do is provide the same throughput as a single threaded KeyDB.

I'm not sure if this will make it for v5.2 but I will definitely want it for the release after.

**Comment by JohnSully** [comment]
The core cause was higher latency at low load.  This has been fixed with change 1a4c3224 which tries to keep at least 50 clients on each thread (configurable with min-clients-per-thread).

The cluster benchmarks were heaviest hit by this because each client has much lower load so the throughput benefits don't have a chance to kick in.  I've verified if you benchmark with memtier using more than the default 4 threads performance is significantly better than Redis or single threaded KeyDB.
